Add moderators option in create_stream_policy, invite_to_stream_policy and stream_post_policy.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR mainly addresses-
- Add moderators only option in `create_stream_policy`.
- Add moderators only option in `invite_to_stream_policy`.
- Add moderators only option in `stream_post_policy`.

This also contains some prep commits for fixing comments in tests, typo in api docs and one for
adding checks for moderators in `test_sending_message_as_stream_post_policy_admins`.

**Testing plan:** <!-- How have you tested? --> Updated existing tests and added new tests.


 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented:
I don't know why Github UI shows commits in this order. The order is correct in my local machine.

timabbott commented:
I merged most of the prep commits as the series ending with fbf5ff26af1d4ae21028381610de455e508ddb8a.  We got a few merge conflicts but I'll comment on details despite them -- just won't rebase locally.

timabbott commented:
These changes look OK aside from the tests needing a rebase.  I think some of the permissions check functions are starting to feel convoluted, though.  For `access_stream_for_send_message` in particular, I wonder if we can should just handle the bots stuff via having a helper that does't consider bots, and then doing e.g.:
```
try: 
    return access_stream_for_send_message_helper(...)
except JsonableError:
    if sender.is_bot and sender.bot_owner is not None:
       return access_stream_for_send_message_helper(...)
```

timabbott commented:
Also might be worth adding an `assert policy_value == STREAM_POST_POLICY_RESTRICT_NEW_MEMBERS` before the last block in `has_permission` just for readability.

sahil839 commented:
@timabbott updated the PR
- Extracted the helper fucntion, not sure whether this is exactly you meant, but I implemented as per my best understanding.
- Added the assert statement as well in `has_permission`.

timabbott commented:
Yeah, that's what I'd had in mind -- it makes the logical structure a lot more clear.  The new semantics are slightly different -- we'll now allow a  `!is_proivisional_member` bot owned by a `is_provisional_member` to post -- which is probably more correct than the opposite, though it's a pretty unlikely situation in the first place.

Merged, thanks @sahil839!

sahil839 commented:
Yes, it was discussed to allow bots to post messages according to the role of their owner [here](https://chat.zulip.org/#narrow/stream/9-issues/topic/stream.20post.20policy.20for.20bots/near/1091346). Also we anyways keep the role of all bots as Member by default and we do not allow to change the role of a bot by UI also.
